README.md
BO (Binary Output)
==================

The Swiss army knife of data examination and manipulation.

This is the tool to use when you need to visualize or convert data in different formats.



How It Works
------------

Think of bo as a data processor that takes in streams of commands and data, and outputs accordingly:

    Command line arguments --\     +------+
                              \    |      |
    Files -----------------------> |  bo  | ---> stdout or file
                              /    |      |
    stdin -------------------/     +------+


Input consists of whitespace separated commands and data, which can be passed in via command line arguments, files, and stdin. Regardless of input source(s), the syntax is exactly the same: whitespace separated commands. Commands determine how input data is to be interpreted, and how output is to be formatted.



Examples
--------

### Brief Introduction

This is a short introduction to understand the examples. See the `Commands` section below for a full description.


#### Commands used in the examples:

  * i{parameters}: Set input type.
  * o{parameters}: Set output type.
  * P{type}: Set the prefix and suffix from a preset ("s" for space separators, and "c" for C-style separators).
  * "a string": Add a string value.

The values i, o, P, and " are used to initiate commands. If bo doesn't recognize a command initiator, it will attempt to interpret numeric data.

#### Input and output type commands consist of multiple fields:

  * **Type**: What type to read data as, or what to output as (see below)
  * **Data Width**: The width of each datum in bytes (1, 2, 4, 8, 16)
  * **Endianness**: What endianness to use when reading or presenting data ("l" or "b" - only required for data width > 1)
  * **Print Width**: Minimum number of digits to use when printing numeric values (only for output. optional).

##### Types:

  * **Integer**: base 16 (h), 10 (i), 8 (o), 2 (b)
  * **Floating point** (f)
  * **String** with c-style escaping (s)
  * **Raw binary data** (B)

For example, `ih2b` means set input type to hexadecimal, 2 bytes per value, big endian. `oo4l11` means set output type to octal, 4 bytes per value, little endian, 11 digits minimum.


#### Guided Example:

    oh1l2 Ps if4l 35.01 10.5 ih1 9f 5a

This stream of commands does the following:

  * set output to hex, 1 byte per value, little endian (only required to separate the last field), min 2 digits per value.
  * Set output preset to "s" (space separated).
  * Set input to float, 4 bytes per value, little endian.
  * Read 35.01 and store internally according to current input type (4 byte little endian float).
  * Read 10.5 and store internally according to current input type (4 byte little endian float).
  * Set input to hex, 1 byte per value (endianness not needed).
  * Read 9f and store internally according to current input type (1 byte hex).
  * Read 5a and store internally according to current input type (1 byte hex).

And output will be `3d 0a 0c 42 00 00 28 41 9f 5a`



### Examples of what you can do with bo

Note: the `-n` flag just appends a newline after processing.


#### See the per-byte layout (_oh1_, _Ps_) of a 4-byte integer (_decimal 12345678_) in big (_ih4b_) or little endian (_ih4l_) format.

    $ bo -n "oh1 Ps ih4b 12345678"
    12 34 56 78

    $ bo -n "oh1 Ps ih4l 12345678"
    78 56 34 12


#### Convert integers between bases.

    $ bo -n "oh8b16 Pc ii8b 1000000"
    0x00000000000f4240

    $ bo -n "oi8b ih8b 7fffffffffffffff"
    9223372036854775807

    $ bo -n "ob8b ih8b ffe846134453a8c2"
    1111111111101000010001100001001101000100010100111010100011000010


#### Examine part of a memory dump (faked in this case) as an array of 32-bit big-endian floats at precision 3.

Fake a 12-byte memory dump as an example:

    $ bo "oB1 if4b 1.1 8.5 305.125" >memory_dump.bin

What it looks like in memory:

    $ bo -n -i memory_dump.bin "oh1b2 Pc iB1"
    0x3f, 0x8c, 0xcc, 0xcd, 0x41, 0x08, 0x00, 0x00, 0x43, 0x98, 0x90, 0x00

What it looks like interpreted as floats:

    $ bo -n -i memory_dump.bin "of4b3 Pc iB1"
    1.100, 8.500, 305.125


#### Convert the endianness of a data dump.

    $ bo "oB1 ih2b 0123 4567 89ab" >data.bin

    $ bo -n -i data.bin "oh2l Pc iB1"
    0x2301, 0x6745, 0xab89

Note: Once input type is set to binary, bo stops parsing, and just passes input DIRECTLY to the output system using the current output type.


#### Convert text files to C-friendly strings.

example.txt:

    This is a test.
    There are tabs  between these words.
    ¡8-Ⅎ⊥∩ sʇɹoddns osʃɐ ʇI

Read in example.txt as raw binary, and output as a C-style escaped string.

    $ bo -n -i example.txt "os iB1"
    This is a test.\nThere are tabs\tbetween\tthese\twords.\n¡8-Ⅎ⊥∩ sʇɹoddns osʃɐ ʇI


#### Reinterpret one type as another at the binary level (for whatever reason).

    $ bo -n "oi4b if4b 1.5"
    1069547520


#### Complex example: Build up a data packet with multiple data types.

    $ bo -n "oh1l2 Pc if4b 1.5 1.25 ii2b 1000 2000 3000 ih1 ff fe 7a ib1 10001011"

Does the following:

  * Set output to hex, 2 ditis per entry (`oh1l2`).
  * Set output prefix/suffix using the "c" preset (`Pc`).
  * Input 1.5 and 1.25 as 4-byte floats, big endian (`if4b 1.5 1.25`).
  * Input 1000, 2000, and 3000 as 2-byte decimal integers, big endian (`ii2b 1000 2000 3000`)
  * Input ff, fe, and 7a as 1-byte hex integers (`ih1 ff fe 7a`)
  * Input 10001011 as 1-byte binary integer (`ib1 10001011`)

Results in:

    0x3f, 0xc0, 0x00, 0x00, 0x3f, 0xa0, 0x00, 0x00, 0x03, 0xe8, 0x07, 0xd0, 0x0b, 0xb8, 0xff, 0xfe, 0x7a, 0x8b



Usage
-----

    bo [options] command [command] ...


### Options

  * -i [filename]: Read commands/data from a file (specifying "-" uses stdin).
  * -o [filename]: Write output to a file (specifying "-" uses stdout).
  * -n Write a newline after processing is complete.
  * -h Print help and exit.
  * -v Print version and exit.


The `bo` command can take input from command line arguments, files (using the -i switch), and stdin (using `-i -`). You may specify as many `-i` switches as you like. Bo first reads all command line arguments, and then reads files in the order they were specified using the `-i` switch.

For example:

    bo -i file1.txt -i - -i file2.txt "oh2b4 Pc" io4b "1 2 3"

Bo will:

  * Parse the string `oh2b4 Pc` (set output type to 2-byte hex, big endian, min 4 digits, then use "c" preset)
  * Parse the string `io4b` (set input type to 4-byte octal, big endian)
  * Parse the string `1 2 3` (read 3 integers)
  * Parse from `file1.txt` (execute all commands contained in the file)
  * Parse from stdin (`-i -`) (execute all commands passed in via stdin)
  * Parse from `file2.txt` (execute all commands contained in the file)

By default, bo outputs to stdout, but you can specify an output file using `-o`.



Commands
--------

Bo parses whitespace separated strings and interprets them as commands. The following commands are supported:

  * i{type}{data width}{endianness}: Specify how to interpret input data
  * o{type}{data width}{endianness}[print width]: Specify how to re-interpret data and how to print it
  * p{string}: Specify a prefix to prepend to each datum output
  * s{string}: Specify a suffix to append to each datum output (except for the last object)
  * P{type}: Specify a preset for prefix and suffix.
  * "...": Read a string value.
  * (numeric): Read a numeric value.

Data is interpreted according to the input format, stored in an intermediary buffer as binary data, and then later re-interpreted and printed according to the output format.

Strings are copied as bytes to the intermediary binary buffer, to be reinterpreted later by the output format.

Before processing any data, both input and output types must be specified via `input type` and `output type` commands. You may later change the input or output types again via commands, even after interpreting data. Any time the output type is changed, all buffers are flushed.



### Input Type Command

The input specifier command consists of 3 parts:

  * Type: How to interpret incoming data (integer, float, decimal, binary, etc)
  * Data Width: The width of each datum in bytes (1, 2, 4, 8, 16)
  * Endianness: What endianness to use when storing data (little or big endian)


### Output Type Command

The output specifier command consists of 4 parts:

  * Type: How to format the data for printing (integer, float, decimal, string, etc)
  * Data Width: The width of each datum in bytes (1, 2, 4, 8, 16)
  * Endianness: What endianness to use when presenting data (little or big endian)
  * Print Width: Minimum number of digits to use when printing numeric values (optional).


#### Type

Determines the type to be used for interpreting incoming data, or presenting outgoing data.

  * Integer (i): Integer in base 10
  * Hexadecimal (h): Integer in base 16
  * Octal (o): Integer in base 8
  * Boolean (b): Integer in base 2
  * Float (f): IEEE 754 binary floating point
  * Decimal (d): IEEE 754 binary decimal (TODO)
  * String (s): String, with c-style encoding for escaped chars (tab, newline, hex, etc).
  * Binary (B): Data is interpreted or output using its binary representation rather than text.

##### Notes on the boolean type


CHANGELOG.md
Change Log
==========


### Version 1.0.1

  * Minor bigfixes
  * Revamped documentation
  * Stricter compiler flags


### Version 1.0.0

  * Initial Release


libbo/src/character_flags.h
/* This file generated by build_character_flags.c */

#ifndef ks_character_flags_H
#define ks_character_flags_H
#ifdef __cplusplus
extern "C" {
#endif


typedef enum
{
    CH_FLAG_NONE         = 0x00,
    CH_FLAG_CONTROL      = 0x01,
    CH_FLAG_WHITESPACE   = 0x02,
    CH_FLAG_BASE_8       = 0x04,
    CH_FLAG_BASE_10      = 0x08,
    CH_FLAG_BASE_16      = 0x10,
    CH_FLAG_FP_NUMBER    = 0x20,
    CH_FLAG_ALPHANUMERIC = 0x40,
    CH_FLAG_PRINTABLE    = 0x80,
} character_flag;

static const unsigned char g_character_flags[256] =
{
    /* 00: NUL   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 01: SOH   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 02: STX   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 03: ETX   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 04: EOT   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 05: ENQ   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 06: ACK   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 07: BEL   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 08: BS    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 09: HT    */ (unsigned char)(CH_FLAG_CONTROL | CH_FLAG_WHITESPACE),
    /* 0a: LF    */ (unsigned char)(CH_FLAG_CONTROL | CH_FLAG_WHITESPACE),
    /* 0b: VT    */ (unsigned char)(CH_FLAG_CONTROL | CH_FLAG_WHITESPACE),
    /* 0c: FF    */ (unsigned char)(CH_FLAG_CONTROL | CH_FLAG_WHITESPACE),
    /* 0d: CR    */ (unsigned char)(CH_FLAG_CONTROL | CH_FLAG_WHITESPACE),
    /* 0e: SO    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 0f: SI    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 10: DLE   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 11: DC1   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 12: DC2   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 13: DC3   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 14: DC4   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 15: NAK   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 16: SYN   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 17: ETB   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 18: CAN   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 19: EM    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 1a: SUB   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 1b: ESC   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 1c: FS    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 1d: GS    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 1e: RS    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 1f: US    */ (unsigned char)(CH_FLAG_CONTROL),
    /* 20: Space */ (unsigned char)(CH_FLAG_WHITESPACE),
    /* 21: !     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 22: "     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 23: #     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 24: $     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 25: %     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 26: &     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 27: '     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 28: (     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 29: )     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 2a: *     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 2b: +     */ (unsigned char)(CH_FLAG_FP_NUMBER | CH_FLAG_PRINTABLE),
    /* 2c: ,     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 2d: -     */ (unsigned char)(CH_FLAG_FP_NUMBER | CH_FLAG_PRINTABLE),
    /* 2e: .     */ (unsigned char)(CH_FLAG_FP_NUMBER | CH_FLAG_PRINTABLE),
    /* 2f: /     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 30: 0     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 31: 1     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 32: 2     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 33: 3     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 34: 4     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 35: 5     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 36: 6     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 37: 7     */ (unsigned char)(CH_FLAG_BASE_8 | CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 38: 8     */ (unsigned char)(CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 39: 9     */ (unsigned char)(CH_FLAG_BASE_10 | CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 3a: :     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 3b: ;     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 3c: <     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 3d: =     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 3e: >     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 3f: ?     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 40: @     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 41: A     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 42: B     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 43: C     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 44: D     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 45: E     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 46: F     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 47: G     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 48: H     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 49: I     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 4a: J     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 4b: K     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 4c: L     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 4d: M     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 4e: N     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 4f: O     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 50: P     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 51: Q     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 52: R     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 53: S     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 54: T     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 55: U     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 56: V     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 57: W     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 58: X     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 59: Y     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 5a: Z     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 5b: [     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 5c: \     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 5d: ]     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 5e: ^     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 5f: _     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 60: `     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 61: a     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 62: b     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 63: c     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 64: d     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 65: e     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_FP_NUMBER | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 66: f     */ (unsigned char)(CH_FLAG_BASE_16 | CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 67: g     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 68: h     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 69: i     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 6a: j     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 6b: k     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 6c: l     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 6d: m     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 6e: n     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 6f: o     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 70: p     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 71: q     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 72: r     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 73: s     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 74: t     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 75: u     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 76: v     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 77: w     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 78: x     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 79: y     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 7a: z     */ (unsigned char)(CH_FLAG_ALPHANUMERIC | CH_FLAG_PRINTABLE),
    /* 7b: {     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 7c: |     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 7d: }     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 7e: ~     */ (unsigned char)(CH_FLAG_PRINTABLE),
    /* 7f: DEL   */ (unsigned char)(CH_FLAG_CONTROL),
    /* 80:       */ (unsigned char)(CH_FLAG_NONE),
    /* 81:       */ (unsigned char)(CH_FLAG_NONE),
    /* 82:       */ (unsigned char)(CH_FLAG_NONE),
    /* 83:       */ (unsigned char)(CH_FLAG_NONE),
    /* 84:       */ (unsigned char)(CH_FLAG_NONE),
    /* 85:       */ (unsigned char)(CH_FLAG_NONE),
    /* 86:       */ (unsigned char)(CH_FLAG_NONE),
    /* 87:       */ (unsigned char)(CH_FLAG_NONE),
    /* 88:       */ (unsigned char)(CH_FLAG_NONE),
    /* 89:       */ (unsigned char)(CH_FLAG_NONE),
    /* 8a:       */ (unsigned char)(CH_FLAG_NONE),
    /* 8b:       */ (unsigned char)(CH_FLAG_NONE),
    /* 8c:       */ (unsigned char)(CH_FLAG_NONE),
    /* 8d:       */ (unsigned char)(CH_FLAG_NONE),
    /* 8e:       */ (unsigned char)(CH_FLAG_NONE),
    /* 8f:       */ (unsigned char)(CH_FLAG_NONE),
    /* 90:       */ (unsigned char)(CH_FLAG_NONE),
    /* 91:       */ (unsigned char)(CH_FLAG_NONE),
    /* 92:       */ (unsigned char)(CH_FLAG_NONE),
    /* 93:       */ (unsigned char)(CH_FLAG_NONE),
    /* 94:       */ (unsigned char)(CH_FLAG_NONE),
    /* 95:       */ (unsigned char)(CH_FLAG_NONE),
    /* 96:       */ (unsigned char)(CH_FLAG_NONE),
    /* 97:       */ (unsigned char)(CH_FLAG_NONE),
    /* 98:       */ (unsigned char)(CH_FLAG_NONE),
    /* 99:       */ (unsigned char)(CH_FLAG_NONE),
    /* 9a:       */ (unsigned char)(CH_FLAG_NONE),
    /* 9b:       */ (unsigned char)(CH_FLAG_NONE),
    /* 9c:       */ (unsigned char)(CH_FLAG_NONE),
    /* 9d:       */ (unsigned char)(CH_FLAG_NONE),
    /* 9e:       */ (unsigned char)(CH_FLAG_NONE),
    /* 9f:       */ (unsigned char)(CH_FLAG_NONE),
    /* a0:       */ (unsigned char)(CH_FLAG_NONE),
    /* a1:       */ (unsigned char)(CH_FLAG_NONE),
    /* a2:       */ (unsigned char)(CH_FLAG_NONE),
    /* a3:       */ (unsigned char)(CH_FLAG_NONE),
    /* a4:       */ (unsigned char)(CH_FLAG_NONE),
    /* a5:       */ (unsigned char)(CH_FLAG_NONE),
    /* a6:       */ (unsigned char)(CH_FLAG_NONE),
    /* a7:       */ (unsigned char)(CH_FLAG_NONE),
    /* a8:       */ (unsigned char)(CH_FLAG_NONE),
    /* a9:       */ (unsigned char)(CH_FLAG_NONE),
    /* aa:       */ (unsigned char)(CH_FLAG_NONE),
    /* ab:       */ (unsigned char)(CH_FLAG_NONE),
    /* ac:       */ (unsigned char)(CH_FLAG_NONE),
    /* ad:       */ (unsigned char)(CH_FLAG_NONE),
    /* ae:       */ (unsigned char)(CH_FLAG_NONE),
    /* af:       */ (unsigned char)(CH_FLAG_NONE),
    /* b0:       */ (unsigned char)(CH_FLAG_NONE),
    /* b1:       */ (unsigned char)(CH_FLAG_NONE),
    /* b2:       */ (unsigned char)(CH_FLAG_NONE),
    /* b3:       */ (unsigned char)(CH_FLAG_NONE),
    /* b4:       */ (unsigned char)(CH_FLAG_NONE),
    /* b5:       */ (unsigned char)(CH_FLAG_NONE),
    /* b6:       */ (unsigned char)(CH_FLAG_NONE),
    /* b7:       */ (unsigned char)(CH_FLAG_NONE),
    /* b8:       */ (unsigned char)(CH_FLAG_NONE),
    /* b9:       */ (unsigned char)(CH_FLAG_NONE),
    /* ba:       */ (unsigned char)(CH_FLAG_NONE),
    /* bb:       */ (unsigned char)(CH_FLAG_NONE),
    /* bc:       */ (unsigned char)(CH_FLAG_NONE),
    /* bd:       */ (unsigned char)(CH_FLAG_NONE),
    /* be:       */ (unsigned char)(CH_FLAG_NONE),
    /* bf:       */ (unsigned char)(CH_FLAG_NONE),
    /* c0:       */ (unsigned char)(CH_FLAG_NONE),
    /* c1:       */ (unsigned char)(CH_FLAG_NONE),
    /* c2:       */ (unsigned char)(CH_FLAG_NONE),
    /* c3:       */ (unsigned char)(CH_FLAG_NONE),
    /* c4:       */ (unsigned char)(CH_FLAG_NONE),
    /* c5:       */ (unsigned char)(CH_FLAG_NONE),
    /* c6:       */ (unsigned char)(CH_FLAG_NONE),
    /* c7:       */ (unsigned char)(CH_FLAG_NONE),
    /* c8:       */ (unsigned char)(CH_FLAG_NONE),
    /* c9:       */ (unsigned char)(CH_FLAG_NONE),
    /* ca:       */ (unsigned char)(CH_FLAG_NONE),
    /* cb:       */ (unsigned char)(CH_FLAG_NONE),
    /* cc:       */ (unsigned char)(CH_FLAG_NONE),
    /* cd:       */ (unsigned char)(CH_FLAG_NONE),
    /* ce:       */ (unsigned char)(CH_FLAG_NONE),
    /* cf:       */ (unsigned char)(CH_FLAG_NONE),
    /* d0:       */ (unsigned char)(CH_FLAG_NONE),
    /* d1:       */ (unsigned char)(CH_FLAG_NONE),
    /* d2:       */ (unsigned char)(CH_FLAG_NONE),
    /* d3:       */ (unsigned char)(CH_FLAG_NONE),
    /* d4:       */ (unsigned char)(CH_FLAG_NONE),
    /* d5:       */ (unsigned char)(CH_FLAG_NONE),
    /* d6:       */ (unsigned char)(CH_FLAG_NONE),
    /* d7:       */ (unsigned char)(CH_FLAG_NONE),
    /* d8:       */ (unsigned char)(CH_FLAG_NONE),
    /* d9:       */ (unsigned char)(CH_FLAG_NONE),
    /* da:       */ (unsigned char)(CH_FLAG_NONE),
    /* db:       */ (unsigned char)(CH_FLAG_NONE),
    /* dc:       */ (unsigned char)(CH_FLAG_NONE),
    /* dd:       */ (unsigned char)(CH_FLAG_NONE),
    /* de:       */ (unsigned char)(CH_FLAG_NONE),
    /* df:       */ (unsigned char)(CH_FLAG_NONE),
    /* e0:       */ (unsigned char)(CH_FLAG_NONE),
    /* e1:       */ (unsigned char)(CH_FLAG_NONE),
    /* e2:       */ (unsigned char)(CH_FLAG_NONE),
    /* e3:       */ (unsigned char)(CH_FLAG_NONE),
    /* e4:       */ (unsigned char)(CH_FLAG_NONE),
    /* e5:       */ (unsigned char)(CH_FLAG_NONE),
    /* e6:       */ (unsigned char)(CH_FLAG_NONE),
    /* e7:       */ (unsigned char)(CH_FLAG_NONE),
    /* e8:       */ (unsigned char)(CH_FLAG_NONE),
    /* e9:       */ (unsigned char)(CH_FLAG_NONE),
    /* ea:       */ (unsigned char)(CH_FLAG_NONE),
    /* eb:       */ (unsigned char)(CH_FLAG_NONE),
    /* ec:       */ (unsigned char)(CH_FLAG_NONE),
    /* ed:       */ (unsigned char)(CH_FLAG_NONE),
    /* ee:       */ (unsigned char)(CH_FLAG_NONE),
    /* ef:       */ (unsigned char)(CH_FLAG_NONE),
    /* f0:       */ (unsigned char)(CH_FLAG_NONE),
    /* f1:       */ (unsigned char)(CH_FLAG_NONE),
    /* f2:       */ (unsigned char)(CH_FLAG_NONE),
    /* f3:       */ (unsigned char)(CH_FLAG_NONE),
    /* f4:       */ (unsigned char)(CH_FLAG_NONE),
    /* f5:       */ (unsigned char)(CH_FLAG_NONE),
    /* f6:       */ (unsigned char)(CH_FLAG_NONE),
    /* f7:       */ (unsigned char)(CH_FLAG_NONE),
    /* f8:       */ (unsigned char)(CH_FLAG_NONE),
    /* f9:       */ (unsigned char)(CH_FLAG_NONE),
    /* fa:       */ (unsigned char)(CH_FLAG_NONE),
    /* fb:       */ (unsigned char)(CH_FLAG_NONE),
    /* fc:       */ (unsigned char)(CH_FLAG_NONE),
    /* fd:       */ (unsigned char)(CH_FLAG_NONE),
    /* fe:       */ (unsigned char)(CH_FLAG_NONE),
    /* ff:       */ (unsigned char)(CH_FLAG_NONE),
};


#ifdef __cplusplus
}
#endif
#endif // ks_character_flags_H
